pkg/vsphere/vm/vm.go

```diff
@@ -731,3 +734,149 @@ func (vm *VirtualMachine) RemoveSnapshotByRef(ctx context.Context, snapshot *typ
 
 	return object.NewTask(vm.Vim25(), res.Returnval), nil
 }
+
+// CheckHostCompatibility checks that a host is compatible for VM placement.
+func (vm *VirtualMachine) CheckHostCompatibility(ctx context.Context, host *types.ManagedObjectReference) (*object.Task, error) {
+	req := types.CheckCompatibility_Task{
+		This: vm.ServiceContent.VmCompatibilityChecker.Reference(),
+		Vm:   vm.Reference(),
+		Host: host,
+	}
+
+	res, err := methods.CheckCompatibility_Task(ctx, vm.Vim25(), &req)
+	if err != nil {
+		return nil, err
+	}
+	return object.NewTask(vm.Vim25(), res.Returnval), nil
+}
+
+// PowerOn powers on a VM. If the environment is VC without DRS enabled, it will attempt to relocate  the VM
+// to the most suitable host in the cluster. If relocation or subsequent power-on fail, it will attempt the next
+// best host, and repeat this process until a successful power-on is achieved or there are no more hosts to try.
+func (vm *VirtualMachine) PowerOn(op trace.Operation) error {
+	drs := vm.Session.DRSEnabled != nil && *vm.Session.DRSEnabled
+	cls := vm.InCluster(op)
+
+	op.Debugf("DRS enabled: %t", drs)
+	op.Debugf("%s resides in a cluster: %t", vm.Reference.String(), cls)
+
+	// otherwise place the VM before powering it on
+	hmp := performance.NewHostMetricsProvider(vm.Session.Vim25())
+	rhp, err := placement.NewRankedHostPolicy(op, vm.Cluster, hmp)
+	if err != nil {
+		return err
+	}
+
+	// any of the following conditions are sufficient to power-on the VM:
+	// - environment is not VC
+	// - DRS is enabled
+	// - the VM does not belong to a cluster
+	// - the current host is adequate for power-on
+	if !vm.IsVC() || drs || !cls || rhp.CheckHost(op, vm.VirtualMachine) {
+		return vm.powerOn(op)
+	}
+
+	op.Warnf("Host is not adequate for power-on, getting placement recommendation")
+
+	var hosts, subset []*object.HostSystem
+
+	f := func() error {
+		hosts, err = rhp.RecommendHost(op, subset)
+		if err != nil {
+			return fmt.Errorf("error recommending host: %s", err.Error())
+		}
+		return nil
+	}
+
+	conf := retry.NewBackoffConfig()
+	conf.MaxInterval = 1 * time.Second
+	conf.MaxElapsedTime = 20 * time.Second
+
+	err = retry.DoWithConfig(f, retry.OnError, conf)
+	if err != nil {
+		return err
+	}
+
+	for len(hosts) > 0 {
+		op.Debugf("hosts: %s", hosts)
+		op.Infof("Placement recommended %s", hosts[0].Reference().String())
+
+		if vm.relocate(op, hosts[0]) != nil {
+			op.Warnf("VM relocation failed: %s", err.Error())
+		} else if vm.powerOn(op) != nil {
+			op.Warnf("VM power-on failed: %s", err.Error())
+		} else {
+			return nil
+		}
+
+		// if relocation or powerOn fails, something has changed and we need a new recommendation
+		subset = hosts[1:]
+
+		if err = retry.DoWithConfig(f, retry.OnError, conf); err != nil {
```

Comment:
Because we are calling this in a loop, if it is called can it potentially return the same host again at some point down the line? You are advancing the `hosts` array by one element. But then you change the `hosts` var by running the retried func again. Is it guaranteed that `RecommendHost` will not return the same host somewhere down the line? Otherwise potential thrashing could occur. especially if you have multiple VCH's on the same system.